Add eager optimizers. Add an MNIST eager training demo (will be unreleased).

nsthorat opened the pull request:
This PR:
* Adds support for eager mode in our optimizers. We reuse the existing optimizer class for backwards compatibility, but add 3 new methods: `minimize`, `computeGradients`, and `applyGradients`. To support eager, optimizers must implement `applyGradients`.
* `optimizers.minimize` takes a function that returns a scalar to be minimized by updating variables.
* `optimizer.computeGradients` by default just calls `math.variableGradients`
* `optimizer.applyGradients` is overridden by each optimizer.
* Add an mnist_eager demo. This is a super simple demo that trains a linear model on MNIST and then shows 50 examples and their predictions.
* Fix `indexToLoc` for scalars and add unit tests.
* Add `util.nextFrame()` which wraps `requestAnimationFrame` in a promise

<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pair-code/deeplearnjs/509)
<!-- Reviewable:end -->

dsmilkov commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> Nice work!

---

Reviewed 23 of 26 files at r1.
Review status: 22 of 24 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[demos/yarn.lock, line 2175 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/509#-L2B2KR2L0OdrlmN8odo:-L2B2KR2L0OdrlmN8odp:b-orpxlr) ([raw file](https://github.com/pair-code/deeplearnjs/blob/90da76e69c81e2db98b91530846be7f4ebd775f0/demos/yarn.lock#L2175)):*
> ../deeplearn-0.4.0.tgz#73e4c6d0f579d9d9252da63a7ba79d5db6fa37f9

publish 0.4.1 and change `demos/package.json`. Then run `yarn` to update this

---

*[demos/mnist_eager/data.ts, line 65 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/509#-L2B5-3KRnhpW1M28TD3:-L2B5-3KRnhpW1M28TD4:bnvanr1) ([raw file](https://github.com/pair-code/deeplearnjs/blob/90da76e69c81e2db98b91530846be7f4ebd775f0/demos/mnist_eager/data.ts#L65)):*
> xs = concatWithNulls(xs, x);

any reason why you need concatWithNulls? Looks like the data loops and will never run out of data. Just seeing if we can reduce amount of demo code.

---

*[demos/mnist_eager/data.ts, line 95 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/509#-L2B4Ct-9tYo5e1nxoPH:-L2B4Ct-9tYo5e1nxoPI:b5wniw8) ([raw file](https://github.com/pair-code/deeplearnjs/blob/90da76e69c81e2db98b91530846be7f4ebd775f0/demos/mnist_eager/data.ts#L95)):*
> ```XML
>   }
> 
>   private getTestData(): dl.NDArray[][] {
> ```

getTestData() is private, but getTrainingData is public. Can we make it private?

---

*[demos/mnist_eager/mnist_eager.ts, line 6 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/509#-L2B66XLsubUsLiCep78:-L2B66XLsubUsLiCep79:bwrt1sd) ([raw file](https://github.com/pair-code/deeplearnjs/blob/90da76e69c81e2db98b91530846be7f4ebd775f0/demos/mnist_eager/mnist_eager.ts#L6)):*
> .05

extract this value as `const learningRate`

---

*[demos/mnist_eager/mnist_eager.ts, line 47 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/509#-L2B6_UGPLo8qwDKDbHl:-L2B6_UGPLo8qwDKDbHm:bb023iu) ([raw file](https://github.com/pair-code/deeplearnjs/blob/90da76e69c81e2db98b91530846be7f4ebd775f0/demos/mnist_eager/mnist_eager.ts#L47)):*
> ```XML
> }
> 
> const statusElement = document.getElementById('status');
> ```

since this demo serves as a tutorial in a way, would be great to have the DLJS related code separate from UI, so people can get a sense of the API. Can we extract the UI code in `mnist_eager/ui.ts` and the dl-related code in `mnist_eager/model.ts`. And the main file can remain `mnist_eager.ts`.

---

*[src/graph_runner.ts, line 18 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/509#-L2B7K3zSlM-KxKDmZvm:-L2B7K3zSlM-KxKDmZvn:b-fpz87q) ([raw file](https://github.com/pair-code/deeplearnjs/blob/90da76e69c81e2db98b91530846be7f4ebd775f0/src/graph_runner.ts#L18)):*
> ```XML
>  */
> 
> import {InputProvider} from './data/input_provider';
> ```

I feel we should migrate `src/graph_runner(_test).ts` to `src/graph/` (feel free to do it here).

---

*[src/util.ts, line 408 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/509#-L2B7v8s3Brjl8Bpcqor:-L2B7v8s3Brjl8Bpcqos:b-jmfyr6) ([raw file](https://github.com/pair-code/deeplearnjs/blob/90da76e69c81e2db98b91530846be7f4ebd775f0/src/util.ts#L408)):*
> () => resolve()

nit: I think you can simplify and fit in one line

`return new Promise<void>(resolve => requestAnimationFrame(resolve));`

---

*[src/math/math.ts, line 52 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/509#-L2B8o-rtwryf8EkMnhm:-L2B8o-rtwryf8EkMnhn:bj8x1pe) ([raw file](https://github.com/pair-code/deeplearnjs/blob/90da76e69c81e2db98b91530846be7f4ebd775f0/src/math/math.ts#L52)):*
> NamedVariableMap = {}

Sorry, I know we've briefly talked about this before, but forgot. Can you remind me what's the reason behind this change?

---

*[src/math/optimizers/optimizer.ts, line 44 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/509#-L2BRf6Py3fI1-T3rnaC:-L2BRf6Py3fI1-T3rnaD:bgrq3ji) ([raw file](https://github.com/pair-code/deeplearnjs/blob/90da76e69c81e2db98b91530846be7f4ebd775f0/src/math/optimizers/optimizer.ts#L44)):*
> variableGradients

expand it here to improve readability:
`const {value, gradients} = this.computeGradients(f);`

---

*[src/math/optimizers/sgd_optimizer.ts, line 79 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/509#-L2BRS1_EvrYvgzFhvTb:-L2BRS1_EvrYvgzFhvTc:br0e94d) ([raw file](https://github.com/pair-code/deeplearnjs/blob/90da76e69c81e2db98b91530846be7f4ebd775f0/src/math/optimizers/sgd_optimizer.ts#L79)):*
> this.c.dispose();

check if c == null before disposing

---

*[src/math/optimizers/sgd_optimizer_test.ts, line 1 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/509#-L2B2tGdkWJ8kBISiP1e:-L2B2tGdkWJ8kBISiP1f:b-z5q2kv) ([raw file](https://github.com/pair-code/deeplearnjs/blob/90da76e69c81e2db98b91530846be7f4ebd775f0/src/math/optimizers/sgd_optimizer_test.ts#L1)):*
> ```XML
> /**
> ```

since the diff is not showing, just double-checking. is this branched from `src/graph/optimizers/sgd_optimizer_test.ts`

---

*[src/math/optimizers/sgd_optimizer_test.ts, line 38 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/509#-L2BSjetUq6e_IMTlKa0:-L2BSjetUq6e_IMTlKa1:b-uv5p9l) ([raw file](https://github.com/pair-code/deeplearnjs/blob/90da76e69c81e2db98b91530846be7f4ebd775f0/src/math/optimizers/sgd_optimizer_test.ts#L38)):*
> ```XML
> 
>     // Cost, internal c, and the new variable value should be the only
>     // additional arrays.
> ```

we are missing an array since `const x = varible(...)` is before `let numArrays = math.getNumArrays();` so it was already accounted for.

Or maybe varible has a bug (my fault). Basically, I would expect this to be true:
```
expect(math.numArrays, 0);
x = variable(Scalar.new(4));
expect(math.numArrays, 1);
x.assign(Scalar.new(5));
expect(math.numArrays, 1); // OLD VALUE SHOULD BE DISPOSED
```

---

*[src/math/optimizers/sgd_optimizer_test.ts, line 55 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/509#-L2BT7pbuSQfxOGikdCH:-L2BT7pbuSQfxOGikdCI:bd764c1) ([raw file](https://github.com/pair-code/deeplearnjs/blob/90da76e69c81e2db98b91530846be7f4ebd775f0/src/math/optimizers/sgd_optimizer_test.ts#L55)):*
> ```XML
>     expect(cost).toBe(null);
> 
>     optimizer.dispose();
> ```

would be nice to dispose the previous cost and check that after optimizer.dispose(), the numArrays is 0

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/509#-:-L2BTwZX-iJeL4QWsasI:b-8m3k00)*
<!-- Sent from Reviewable.io -->

nsthorat commented:
Review status: 13 of 29 files reviewed at latest revision, 13 unresolved discussions, some commit checks pending.

---

*[demos/yarn.lock, line 2175 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/509#-L2B2KR2L0OdrlmN8odo:-L2BmTnx8ik7rj4H-9gs:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/90da76e69c81e2db98b91530846be7f4ebd775f0/demos/yarn.lock#L2175)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

> ../deeplearn-0.4.0.tgz#73e4c6d0f579d9d9252da63a7ba79d5db6fa37f9

publish 0.4.1 and change `demos/package.json`. Then run `yarn` to update this
</blockquote></details>

Done.

---

*[demos/mnist_eager/data.ts, line 65 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/509#-L2B5-3KRnhpW1M28TD3:-L2BVG0t4KC98jZoELfL:bkxkx8j) ([raw file](https://github.com/pair-code/deeplearnjs/blob/90da76e69c81e2db98b91530846be7f4ebd775f0/demos/mnist_eager/data.ts#L65)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

> xs = concatWithNulls(xs, x);

any reason why you need concatWithNulls? Looks like the data loops and will never run out of data. Just seeing if we can reduce amount of demo code.
</blockquote></details>

This is to work around the lack of "stack":
https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.stack.html

I added a TODO over this method. Basically, I think it would be nice if our goal was to try to simplify this demo as much as humanly possible. This will give us a good sense for what users will typically see.

---

*[demos/mnist_eager/data.ts, line 95 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/509#-L2B4Ct-9tYo5e1nxoPH:-L2BbFbvlr8hgBiAJhpN:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/90da76e69c81e2db98b91530846be7f4ebd775f0/demos/mnist_eager/data.ts#L95)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

getTestData() is private, but getTrainingData is public. Can we make it private?
</blockquote></details>

Done.

---

*[demos/mnist_eager/mnist_eager.ts, line 6 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/509#-L2B66XLsubUsLiCep78:-L2BmRfrPI7w-u1XZWz_:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/90da76e69c81e2db98b91530846be7f4ebd775f0/demos/mnist_eager/mnist_eager.ts#L6)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

> .05

extract this value as `const learningRate`
</blockquote></details>

Done.

---

*[demos/mnist_eager/mnist_eager.ts, line 47 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/509#-L2B6_UGPLo8qwDKDbHl:-L2BmS-Kg2WZRLKM1LmM:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/90da76e69c81e2db98b91530846be7f4ebd775f0/demos/mnist_eager/mnist_eager.ts#L47)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

since this demo serves as a tutorial in a way, would be great to have the DLJS related code separate from UI, so people can get a sense of the API. Can we extract the UI code in `mnist_eager/ui.ts` and the dl-related code in `mnist_eager/model.ts`. And the main file can remain `mnist_eager.ts`.
</blockquote></details>

Done.

---

*[src/util.ts, line 408 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/509#-L2B7v8s3Brjl8Bpcqor:-L2BW4OvTGXT1bqbhUgI:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/90da76e69c81e2db98b91530846be7f4ebd775f0/src/util.ts#L408)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

> () => resolve()

nit: I think you can simplify and fit in one line

`return new Promise<void>(resolve => requestAnimationFrame(resolve));`
</blockquote></details>

Done.

---

*[src/math/math.ts, line 52 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/509#-L2B8o-rtwryf8EkMnhm:-L2BWBHTM2mw_zbmQEqi:b-syxygt) ([raw file](https://github.com/pair-code/deeplearnjs/blob/90da76e69c81e2db98b91530846be7f4ebd775f0/src/math/math.ts#L52)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

> NamedVariableMap = {}

Sorry, I know we've briefly talked about this before, but forgot. Can you remind me what's the reason behind this change?
</blockquote></details>

This is so we interoperate with valueAndGradients (we pass registered variables directly to variableGradients).

---

*[src/math/optimizers/optimizer.ts, line 44 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/509#-L2BRf6Py3fI1-T3rnaC:-L2BWRsZvesul8AUCC9c:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/90da76e69c81e2db98b91530846be7f4ebd775f0/src/math/optimizers/optimizer.ts#L44)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

> variableGradients

expand it here to improve readability:
`const {value, gradients} = this.computeGradients(f);`
</blockquote></details>

Done.

---

*[src/math/optimizers/sgd_optimizer.ts, line 79 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/509#-L2BRS1_EvrYvgzFhvTb:-L2BWbIV1AJFs2RJMQWk:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/90da76e69c81e2db98b91530846be7f4ebd775f0/src/math/optimizers/sgd_optimizer.ts#L79)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

> this.c.dispose();

check if c == null before disposing
</blockquote></details>

Done.

---

*[src/math/optimizers/sgd_optimizer_test.ts, line 1 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/509#-L2B2tGdkWJ8kBISiP1e:-L2BWcXOnI2XkiGGPJ7D:b-15tlr5) ([raw file](https://github.com/pair-code/deeplearnjs/blob/90da76e69c81e2db98b91530846be7f4ebd775f0/src/math/optimizers/sgd_optimizer_test.ts#L1)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

since the diff is not showing, just double-checking. is this branched from `src/graph/optimizers/sgd_optimizer_test.ts`
</blockquote></details>

Yes, apologies, nothing changed except I added the new test.

---

*[src/math/optimizers/sgd_optimizer_test.ts, line 38 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/509#-L2BSjetUq6e_IMTlKa0:-L2BYWW9hTy12_qwcYYL:b-e10le1) ([raw file](https://github.com/pair-code/deeplearnjs/blob/90da76e69c81e2db98b91530846be7f4ebd775f0/src/math/optimizers/sgd_optimizer_test.ts#L38)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

we are missing an array since `const x = varible(...)` is before `let numArrays = math.getNumArrays();` so it was already accounted for.

Or maybe varible has a bug (my fault). Basically, I would expect this to be true:
```
expect(math.numArrays, 0);
x = variable(Scalar.new(4));
expect(math.numArrays, 1);
x.assign(Scalar.new(5));
expect(math.numArrays, 1); // OLD VALUE SHOULD BE DISPOSED
```
</blockquote></details>

Looks like a bug in variable assign. 302 your PR

---

*[src/math/optimizers/sgd_optimizer_test.ts, line 55 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/509#-L2BT7pbuSQfxOGikdCH:-L2BZj6vFWLy9WJ6upHs:buw9bt7) ([raw file](https://github.com/pair-code/deeplearnjs/blob/90da76e69c81e2db98b91530846be7f4ebd775f0/src/math/optimizers/sgd_optimizer_test.ts#L55)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

would be nice to dispose the previous cost and check that after optimizer.dispose(), the numArrays is 0
</blockquote></details>

There are 3 left because of the bug with variable

---

*[src/graph_runner.ts, line 18 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/509#-L2B7K3zSlM-KxKDmZvm:-L2BVvWA9B-e1_ztH-A-:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/90da76e69c81e2db98b91530846be7f4ebd775f0/src/graph_runner.ts#L18)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

I feel we should migrate `src/graph_runner(_test).ts` to `src/graph/` (feel free to do it here).
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/509)*
<!-- Sent from Reviewable.io -->
